text/0003-placeholders-in-conditions.md

```diff
@@ -0,0 +1,145 @@
+
+Link-time validation has three drawbacks, though.
+
+1. Fundamental errors in a template are not discovered until the template is first linked. E.g., `permit(principal in ?principal,action,resource) when { 1 < "hello" >}` will _always_ be invalid when linked, and it would be nice to know that as early as possible.
+
+2. It may be more difficult to analyze the possible impact of a template by considering all of its possible safe linkages. Providing the types essentially puts useful bounds on how templates can be linked safely in the future.
+
+3. Validation is more expensive at linkage time. If we provided types for placeholders, we could mostly validate the template when it is added to the store, and then just check at link time that the linkages to variables have the correct types.
+
+# Drawbacks
+
+This adds new syntax to the Cedar language.
+
+It relegates validation to link time, which has the drawbacks mentioned in the previous section.
+
+# Alternatives
+
+## Type annotations
+
+One obvious alternative would be to annotate placeholder variables with types so that templates can be validated without knowing particular linkages. For example, we might rewrite the example above to be:
+```
+template (?user:User, ?resourcebound:DeviceGroup)
+permit(principal, action, resource)
+when { 
+     ?resourcebound in resource 
+  && principal in ?user.delegates
+};
+```
+Here we have added `:User` and `:DeviceGroup` type annotations to the introduced variables. The validator can treat `?user` and `?resourcebound` as respectively having these types when considering this template, e.g., to make sure that `?user` truly does have a `delegates` attribute.
+
+We could also allow the appearance of `?principal` and `?resource` in the list of variables, in order to give them types as well. Doing so is probably not necessary, however, because the types of these placeholder variables can be inferred from the schema (based on entity type given to `principal` and `resource` during "cross product" validation, and respectively the `memberOfTypes` portions of those entity types).
+
+This alternative is more work to implement and more effort for users, but mitigates the drawbacks mentioned above.
+
+## Use type parameter syntax
+
+Rather than write `template (?user, ?resourcebound) permit ( principal, ...)` an alternative would be to write `permit<?user,?resourcebound>( principal, ...)`. Using the `<...>` notation resembles parameterization on types used in other languages, as with C++ templates and Java generics. 
+
+The benefit of this syntax is that (1) it does not introduce a new keyword (`template`); (2) a normal policy looks closer to a "template" without parameters; and (3) it follows conventions similar to previous languages. I don't know if the parser would be any harder to write with this alternative syntax.
```

Comment:
I slightly prefer this alternative syntax. I feel like `template(...)` will be confused with `@template(...)` since prior to this proposed change, annotations were the only construct allowed before `permit` or `forbid`.

Of course, we could have special error messages to help guide users in that case. But it still feels a little ugly to me to add a new keyword.

Reply:
I made a comment above which I should have made inline on this text.  Adding it here:

I think the `permit<?user,?resourcebound>` syntax, while it might be more familiar to C++ or Java programmers, is less readable for less-technically-savvy users, so I'm against it. One of Cedar's goals is to be readable for non-experts and non-programmers.

Reply:
I also don't quite understand the worry about `template(...)` being confused with `@template(...)`.  Is it any different than possible confusion between `permit(...)` and `@permit(...)`, which could happen today?

Reply:
Fair point about `@permit(...)`.  Also a fair point that our target audience may not be familiar with templates in other languages -- I may also be biasing towards `permit<?user,?resourcebound>` due to C++ experience.